### PR TITLE
Add a body property to API Gateway RestAPI for Swagger import support

### PR DESCRIPTION
**Pull request by lukehoban**
## Description

The `body` property allows supplying a Swagger/OpenAPI spec to a RestAPI to create the API definition using the routes and integrations defined in the spec.

Related to https://github.com/hashicorp/terraform/issues/11034 and the work done in https://github.com/hashicorp/terraform/pull/6175.

Provides similar behaviour to what's available in CloudFormation with: http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-apigateway-restapi.html#cfn-apigateway-restapi-body.

## Tests
```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSAPIGatewayRestApi_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSAPIGatewayRestApi_ -timeout 120m
=== RUN   TestAccAWSAPIGatewayRestApi_basic
--- PASS: TestAccAWSAPIGatewayRestApi_basic (50.66s)
=== RUN   TestAccAWSAPIGatewayRestApi_openapi
--- PASS: TestAccAWSAPIGatewayRestApi_openapi (28.66s)
PASS
ok  	github.com/pulumi/terraform-provider-aws/aws	79.354s
```

### CODE REVIEW

**Comment by lukehoban** [comment]
@radeksimko Please let me know if there's any questions on concerns on this that I can help address.

**Comment by lukehoban** [comment]
Updated tests to create valid REST API definitions and add test results to summary.

cc @radeksimko @Ninir @betabandido @apparentlymart for feedback.

File: aws/resource_aws_api_gateway_rest_api.go [hunk 1]
```diff
@@ -76,6 +81,17 @@ func resourceAwsApiGatewayRestApiCreate(d *schema.ResourceData, meta interface{}
...
 	d.SetId(*gateway.Id)
 
+	if body, ok := d.GetOk("body"); ok {
+		_, err := conn.PutRestApi(&apigateway.PutRestApiInput{
+			RestApiId: gateway.Id,
+			Mode:      aws.String("overwrite"),
+			Body:      []byte(body.(string)),
+		})
+		if err != nil {
+			return fmt.Errorf("Error putting API Gateway specification: %s", err)
```
Comment (Ninir) [inline aws/resource_aws_api_gateway_rest_api.go]: Could you use `errwrap.Wrapf("Error creating API Gateway specification: {{err}}", err)` there please? (notice the creating word, to make it clear we are putting by creating)

File: aws/resource_aws_api_gateway_rest_api.go [hunk 2]
```diff
@@ -76,6 +81,17 @@ func resourceAwsApiGatewayRestApiCreate(d *schema.ResourceData, meta interface{}
 
 	d.SetId(*gateway.Id)
 
+	if body, ok := d.GetOk("body"); ok {
+		_, err := conn.PutRestApi(&apigateway.PutRestApiInput{
+			RestApiId: gateway.Id,
+			Mode:      aws.String("overwrite"),
```
Comment (Ninir) [inline aws/resource_aws_api_gateway_rest_api.go]: Can you use the `apigateway.PutModeOverwrite` constant here and line 178?

File: aws/resource_aws_api_gateway_rest_api_test.go [hunk 3]
```diff
@@ -144,3 +216,81 @@ resource "aws_api_gateway_rest_api" "test" {
...
+              "statusCode": 200
+            }
+          }
+        }
+      }
+    }
+  }
+}
+EOF
+}
```
Comment (Ninir) [inline aws/resource_aws_api_gateway_rest_api_test.go]: Can you fix the space indentation here please? 😄 (mostly at the beginning of the JSON structure)

File: aws/resource_aws_api_gateway_rest_api.go [hunk 4]
```diff
@@ -155,6 +171,19 @@ func resourceAwsApiGatewayRestApiUpdate(d *schema.ResourceData, meta interface{}
...
 
+	if d.HasChange("body") {
+		if body, ok := d.GetOk("body"); ok {
+			_, err := conn.PutRestApi(&apigateway.PutRestApiInput{
+				RestApiId: aws.String(d.Id()),
+				Mode:      aws.String("overwrite"),
+				Body:      []byte(body.(string)),
+			})
+			if err != nil {
+				return err
```
Comment (Ninir) [inline aws/resource_aws_api_gateway_rest_api.go]: I would go for the same as line 91 here: `errwrap.Wrapf("Error updating API Gateway specification: {{err}}", err)`. Thoughts?

File: aws/resource_aws_api_gateway_rest_api.go [hunk 5]
```diff
@@ -76,6 +81,17 @@ func resourceAwsApiGatewayRestApiCreate(d *schema.ResourceData, meta interface{}
 
 	d.SetId(*gateway.Id)
 
+	if body, ok := d.GetOk("body"); ok {
+		_, err := conn.PutRestApi(&apigateway.PutRestApiInput{
```
Comment (Ninir) [inline aws/resource_aws_api_gateway_rest_api.go]: Could you log in the logs that we are setting the specification?
Just for verbosity & debugging :D

Same for the update would be highly appreciated!

File: aws/resource_aws_api_gateway_rest_api.go (hunk 1 above)
Comment (radeksimko) [inline aws/resource_aws_api_gateway_rest_api.go]: It does no harm, but I think the wrapper here isn't necessary as the only place the error will escalate to is the UI, where all that matters is string-based error message user will read, not the original error type.

File: aws/resource_aws_api_gateway_rest_api.go (hunk 4 above)
Comment (radeksimko) [inline aws/resource_aws_api_gateway_rest_api.go]: It does no harm, but I think the wrapper here isn't necessary as the only place the error will escalate to is the UI, where all that matters is string-based error message user will read, not the original error type.

**Comment by radeksimko** [comment]
Thanks @Ninir - other than what you mentioned we should also document clearly which resources collide with this field/approach, I believe these are:

 - `aws_api_gateway_gateway_response`
 - `aws_api_gateway_integration`
 - `aws_api_gateway_integration_response`
 - `aws_api_gateway_method`
 - `aws_api_gateway_method_response`
 - `aws_api_gateway_method_settings`
 - `aws_api_gateway_model`
 - `aws_api_gateway_resource`

i.e. it should be clear to the user that they cannot use both Swagger and these resources together for the same API.

**Comment by lukehoban** [comment]
Thanks @ninir and @radeksimko!  

Updated PR to address feedback.  Please take a look at the doc updates in particular.

File: website/docs/r/api_gateway_rest_api.html.markdown [hunk 6]
```diff
@@ -26,6 +26,18 @@ The following arguments are supported:
 * `name` - (Required) The name of the REST API
 * `description` - (Optional) The description of the REST API
 * `binary_media_types` - (Optional) The list of binary media types supported by the RestApi. By default, the RestApi supports only UTF-8-encoded text payloads.
+* `body` - (Optional) An OpenAPI specification that defines the set of routes and integrations to create as part of the REST API.
+
+__Note__: If the `body` argument is provided, the OpenAPI specification will be used to configure the resources, methods and integrations for the Rest API. If this argument is provided, the following resources should not be configured manually for the same REST API, as updates may cause manual resource updates to be overwritten:
```
Comment (radeksimko) [inline website/docs/r/api_gateway_rest_api.html.markdown]: > the following resources should not be configured manually

I'd rather say _should not be managed as separate ones_ as that's more in-line with the language used in the rest of our docs.

**Comment by Ninir** [comment]
hey @lukehoban !

Just pushed the documentation update following @radeksimko 's recommandation.
Thanks for all the work here, looks very good! :)

Cheers!

**Comment by kirangangineni** [comment]
Hi All - When will this feature be available. I have v0.10.5 on my mac. 

resource "aws_api_gateway_rest_api" "MyDemoAPI" {
  name        = "MyDemoAPI"
  description = "This is my API for demonstration purposes"
  body = "./api/swagger.yml"
}

when I run terraform plan , I get 
1 error(s) occurred:

* module.apigateway.aws_api_gateway_rest_api.MyDemoAPI: : invalid or unknown key: body

Am I missing some thing here. Please let me know

**Comment by bvingerh** [comment]
Same with v0.10.6: body argument not recognized.
That's the fourth release of Terraform without a release of the AWS provider (which is stuck at v0.1.4 of August 8)...

**Comment by kirangangineni** [comment]
I built the terraform-provider-aws using the instructions from README. Now when I run terraform plan, the body tag is recognized. But when I apply, I get the following error. I used the same swagger with AWS SDK/CLI and it worked fine. Can you please help

$ terraform apply
module.apigateway.aws_api_gateway_rest_api.MyDemoAPI: Creating...
  body:             "" => "../api/swagger.json"
  created_date:     "" => "<computed>"
  description:      "" => "This is my API for demonstration purposes"
  name:             "" => "MyDemoAPI"
  root_resource_id: "" => "<computed>"
Error applying plan:

1 error(s) occurred:

* module.apigateway.aws_api_gateway_rest_api.MyDemoAPI: 1 error(s) occurred:

* aws_api_gateway_rest_api.MyDemoAPI: Error creating API Gateway specification: BadRequestException: Invalid Swagger 2.0 input.
	status code: 400, request id: 5ec75ec4-a235-11e7-b1f5-7b2e853b338e

**Comment by woody3549** [comment]
Hi all,

Thanks for this great work.
Can you please let me know when it's going to be released ?

Thanks

**Comment by ebarault** [comment]
i have the same error as @kirangangineni, with the same difference that i'm using a yaml swagger file. Are we supposed to provide the file in a special manner?

**Comment by Ninir** [comment]
Hi folks,

the AWS provider (this repository) is release independently from the core one (https://github.com/hashicorp/terraform).
This means you can use Terraform 0.10.8 with the AWS Provider version 1.0.0, 1.1.0, etc..

the body property was added in the 1.0.0 release: to ensure you get it right, remove the `.terraform/plugins` directory and run `terraform init` again: this will get the latest version and you should be all set.

Thus being said, @ebarault is right: the body property expects the body of the file, not the path to the file. As in:

# using data template
```hcl
resource "aws_api_gateway_rest_api" "api" {
  ...
  body = "${data.template_file.swagger_api.rendered}"
}
```

# using a file
```hcl
resource "aws_api_gateway_rest_api" "api" {
  ...
  body = "${file("myfile.json"}"
}
```

# using inline strings
```hcl
resource "aws_api_gateway_rest_api" "api" {
  ...
  body = <<YAML
myfile
YAML
}
```

Tell us how it goes! :)

**Comment by Willis0826** [comment]
@Ninir 
Thanks for the clearly demonstration : ) But here is a typo in your post :
> body = "${file("myfile.json"}"
I believe it should be ${file("...")}

**Pull request merged**